kapitan/refs/secrets/vaultkv.py

```diff
@@ -0,0 +1,285 @@
+    * VAULT_TOKEN: token for vault
+    * VAULT_ROLE_ID: (required by approle)
+    * VAULT_SECRET_ID: (required by approle)
+    * VAULT_USERNAME: username to login to vault
+    * VAULT_PASSWORD: password to login to vault
+    * VAULT_CLIENT_KEY: path to an unencrypted PEM-encoded private key matching the client certificate
+    * VAULT_CLIENT_CERT: path to a PEM-encoded client certificate for TLS authentication to the Vault server
+    * VAULT_CACERT: path to a PEM-encoded CA cert file to use to verify the Vault server TLS certificate
+    * VAULT_CAPATH: path to a directory of PEM-encoded CA cert files to verify the Vault server TLS certificate
+    * VAULT_NAMESPACE: specify the Vault Namespace, if you have one
+    """
+    # Helper lambda function to get varibles either from parameters or environment
+    get_variable = lambda x: parameter.get(x, os.getenv('VAULT_'+x.upper(),default=''))
+    env = {}
+    env['url'] = get_variable('addr')
+    env['namespace'] = get_variable('namespace')
+    # VERIFY VAULT SERVER TLS CERTIFICATE
+    skip_verify = get_variable('skip_verify')
+    if skip_verify.lower() == 'false':
+        cert = get_variable('cacert')
+        if not cert:
+            cert_path = get_variable('capath')
+            if not cert_path:
+                raise Exception('Neither VAULT_CACERT nor VAULT_CAPATH specified')
+            env['verify'] = cert_path
+        else:
+            env['verify'] = cert
+
+    else:
+        env['verify'] = False
+    # CLIENT CERTIFICATE FOR TLS AUTHENTICATION
+    client_key = get_variable('client_key')
+    client_cert = get_variable('client_cert') 
+    if client_key != '' and client_cert != '':
+        env['cert'] = (client_cert,client_key)
+    return env
+
+def vault_obj(vault_parameters):
+    """
+    vault_parameters: necessary parameters to authenticate & get value from vault, provided by inventory
+    e.g.:
+        addr: http://127.0.0.1:8200
+        auth: userpass
+        client_cert: /path/to/cert
+        client_key: /path/to/key
+        namespace: CICD-alpha
+        skip_verify: false
+    Authenticate client to server and return client object
+    """
+    env = get_env(vault_parameters)
+
+    client = hvac.Client(
+        **{k:v for k,v in env.items() if k not in ('auth', None)}
+    )
+
+    auth_type = vault_parameters['auth']
+    # GET TOKEN EITHER FROM ENVIRONMENT OF FILE
+    if auth_type in ['token','github']:
+        env['token'] = os.getenv('VAULT_TOKEN')
+        if not env['token']:
+            try:
+                with open(os.path.join(os.path.expanduser('~'),'.vault-token'),'r') as f:
+                    env['token'] = f.read()
+            except IOError:
+                VaultError("Cannot read file ~/.vault-token")
+    # DIFFERENT LOGIN METHOD BASED ON AUTHENTICATION TYPE
+    if auth_type == 'token':
+        client.token = env['token']
+    elif auth_type == 'ldap':
+        client.auth.ldap.login(
```

Comment:
This needs a check to see if both env vars here exist before using them (?). 
If they don't exist log an info message?

Reply:
There is a `client.is_authenticated()`at the end of the function that I expected to handle the authentication failure.